ariadne/graphql.py

```diff
@@ -17,48 +37,73 @@
     root_value: Optional[RootValue] = None,
     debug: bool = False,
     logger: Optional[str] = None,
-    validation_rules=None,
+    validation_rules: Optional[Sequence[RuleType]] = None,
     error_formatter: ErrorFormatter = format_error,
-    middleware: Middleware = None,
+    middleware: Optional[MiddlewareManager] = None,
+    extensions: Optional[List[Type[Extension]]] = None,
     **kwargs,
 ) -> GraphQLResult:
-    try:
-        validate_data(data)
-        query, variables, operation_name = (
-            data["query"],
-            data.get("variables"),
-            data.get("operationName"),
-        )
-        document = parse(query)
+    extension_manager = ExtensionManager(extensions)
 
-        if validation_rules:
-            errors = _graphql.validate(schema, document, validation_rules)
-            if errors:
-                return handle_graphql_errors(
-                    errors, logger=logger, error_formatter=error_formatter, debug=debug
-                )
+    with extension_manager.request(context_value):
+        try:
+            validate_data(data)
+            query, variables, operation_name = (
+                data["query"],
+                data.get("variables"),
+                data.get("operationName"),
+            )
 
-        if callable(root_value):
-            root_value = root_value(context_value, document)
+            with extension_manager.parsing(query):
+                document = parse_query(query)
 
-        result = await _graphql.graphql(
-            schema,
-            query,
-            root_value=root_value,
-            context_value=context_value,
-            variable_values=variables,
-            operation_name=operation_name,
-            middleware=middleware,
-            **kwargs,
-        )
-    except GraphQLError as error:
-        return handle_graphql_errors(
-            [error], logger=logger, error_formatter=error_formatter, debug=debug
-        )
-    else:
-        return handle_query_result(
-            result, logger=logger, error_formatter=error_formatter, debug=debug
-        )
+            with extension_manager.validation(context_value):
```

Comment:
Should this be `validating(...)` (to match `parsing(...)`)?

Reply:
Those are named after [`graphql-extensions`](https://github.com/apollographql/apollo-server/blob/master/packages/graphql-extensions/src/index.ts#L102) which does `parsing`, `validation` and `execution`, but seeing how this API's not publish in Apollo Server, we may go for something else. I am not sure if I like `def validating` call `validation_started` & `validation_finished`